gquant/dataframe_flow/node.py

```diff
@@ -30,34 +34,84 @@ def __init__(self, task):
         self.outputs = []
         self.visited = False
         self.input_df = {}
+
         self.input_columns = {}
+        # input_columns format:
+        #     {
+        #         to_port1: {
+        #             col1_name: col1_type,
+        #             col2_name: col2_type,
+        #             ... etc.
+        #         },
+        #         to_port2: { ... }
+        #         ... etc.
+        #     }
+
         self.output_columns = {}
+
         self.clear_input = True
-        self.required = None
-        self.addition = None
-        self.deletion = None
+        self.required = {}
+        self.addition = {}
+        self.deletion = {}
+        # Retention must be None instead of empty dict. This replaces anything
+        # set by required/addition/retention. An empty dict is a valid setting
+        # for retention therefore use None instead of empty dict.
         self.retention = None
-        self.rename = None
+        self.rename = {}
         self.delayed_process = False
         # customized the column setup
         self.columns_setup()
 
+    def ports_setup(self):
+        """Virtual method for specifying inputs/outputs ports.
+        """
+        raise NotImplementedError
+
+    def using_ports(self):
+        try:
+            _ = self.ports_setup()
+            has_ports = True
+        except NotImplementedError:
+            has_ports = False
+        return has_ports
+
+    def __get_io_port(self, io=None, full_port_spec=False):
+        input_ports, output_ports = self.ports_setup()
+        if io in ('in',):
+            io_ports = input_ports
+        else:
+            io_ports = output_ports
+
+        if io_ports is None:
+            io_ports = dict()
+
+        if not full_port_spec:
+            io_ports = list(io_ports.keys())
+
+        return io_ports
+
+    def get_input_ports(self, full_port_spec=False):
+        return self.__get_io_port(io='in', full_port_spec=full_port_spec)
+
+    def get_output_ports(self, full_port_spec=False):
+        return self.__get_io_port(io='out', full_port_spec=full_port_spec)
+
     def __translate_column(self, columns):
         output = {}
-        for k in columns:
-            types = columns[k]
-            if types is not None and types.startswith("@"):
-                types = self.conf[types[1:]]
-            if k.startswith("@"):
-                field_name = k[1:]
+        for col_name, col_type in columns.items():
+            if col_type is not None and col_type.startswith("@"):
```

Comment:
The `translate_column` is useful but it could be factored out to just be a node helper function and we can add examples of how to use that in `columns_setup` API. People can then use it or implement their own logic for dynamic columns manipulation. By factoring it out we wouldn't be coupling a mini language (DSL - domain-specific lang.) to the node class to manipulate columns.

Reply:
that's a good idea to make translate_column a customizable thing.

Reply:
I think this is a TODO item. Add an issue to github. I don't want to refactor existing nodes in this PR, it would be too much and clutter up the PR.

If we end up merging and using ports we can modify/add nodes afterwards. I don't want to break existing notebooks and workflows/taskgraphs. These can be addressed separately.